File: sdk/go/serde.go
// Package serde implements Confluent Schema Registry-aware serialization and
// deserialization of proto.Message values for Kafka.
//
// Design (Decision A — "thin SDK"):
//   - The SDK NEVER reads a .proto file and NEVER registers schemas. Schema
//     registration is an out-of-band step (CI/ops in the contracts repo).
//   - At Bind, the SDK RESOLVES the schema_id for a topic from the Schema
//     Registry (it only reads). Produce stamps that id; Consume validates the
//     id against the topic's subject and deserializes into the bound type.
//
// Wire format (Confluent SR):
//
//	[0x00 magic] [schema_id: 4 bytes BE] [message-index array] [proto3 payload]
//
// The message-index array is variable length (Confluent encoding): a single
// top-level message at index 0 is the 1-byte optimization 0x00; otherwise it is
// zig-zag varints (count, then each index). Subject naming uses
// TopicNameStrategy: "<topic>-value".
//
// SECURITY: Kafka accepts arbitrary bytes; this SDK does NOT gate writes at the
// broker. Enforcement lives at the CONSUMER: Consume rejects (typed errors, for
// DLQ routing) anything that is not a valid envelope carrying a schema_id that
// is a registered version of THIS topic's subject and that deserializes into the
// bound type. Sender authentication (ACLs/TLS/signing) is a separate layer.
package serde

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sync"
	"time"

	"google.golang.org/protobuf/proto"
)

const magicByte = byte(0x00)

// Sentinel errors. Consume wraps these so adapters can route to DLQ via errors.Is.
var (
	ErrInvalidMagicByte   = errors.New("serde: invalid magic byte: expected 0x00")
	ErrFrameTooShort      = errors.New("serde: frame too short")
	ErrTopicNotBound      = errors.New("serde: topic not bound; call Bind before Produce/Consume")
	ErrSchemaForeign      = errors.New("serde: schema_id is not a registered version of this topic's subject")
	ErrMessageIndexMismatch = errors.New("serde: message-index does not match the bound type")
	ErrDeserialize        = errors.New("serde: protobuf deserialization failed")
)

// Config holds Schema Registry connection settings.
type Config struct {
	SRURL     string // SCHEMA_REGISTRY_URL (use https for TLS)
	APIKey    string // SCHEMA_REGISTRY_API_KEY  (optional)
	APISecret string // SCHEMA_REGISTRY_API_SECRET (optional)
}

// binding holds everything resolved for one topic at Bind time.
type binding struct {
	prototype     proto.Message
	subject       string
	schemaID      uint32 // resolved from SR (latest version of subject); used by Produce
	msgIndexBytes []byte // pre-encoded message-index for Produce
	indexes       []int  // expected message-index path, for Consume validation
}

// Serde resolves schema ids from the Schema Registry and frames/unframes
// Confluent envelopes. Construct with New or NewWithConfig.
type Serde struct {
	cfg    Config
	client *http.Client

	mu       sync.RWMutex
	bindings map[string]binding             // topic → binding
	idOK     map[string]struct{}            // consume cache: "<id>|<subject>" confirmed
}

// New creates a Serde reading SR config from environment variables:
//   - SCHEMA_REGISTRY_URL (required)
//   - SCHEMA_REGISTRY_API_KEY / SCHEMA_REGISTRY_API_SECRET (optional)
func New() (*Serde, error) {
	srURL := os.Getenv("SCHEMA_REGISTRY_URL")
	if srURL == "" {
		return nil, fmt.Errorf("serde: SCHEMA_REGISTRY_URL environment variable is not set")
	}
	return NewWithConfig(Config{
		SRURL:     srURL,
		APIKey:    os.Getenv("SCHEMA_REGISTRY_API_KEY"),
		APISecret: os.Getenv("SCHEMA_REGISTRY_API_SECRET"),
	}), nil
}

// NewWithConfig creates a Serde with an explicit Config.
func NewWithConfig(cfg Config) *Serde {
	return &Serde{
		cfg:      cfg,
		client:   &http.Client{Timeout: 10 * time.Second},
		bindings: make(map[string]binding),
		idOK:     make(map[string]struct{}),
	}
}

// Bind maps a topic to its proto message type and resolves the topic's schema_id
// from the Schema Registry (subject "<topic>-value", latest version). It reads
// only — it never registers. Fails fast if the subject is not registered yet.
func (s *Serde) Bind(topic string, prototype proto.Message) error {
	subject := topicSubject(topic)
	id, err := s.srLatestID(subject)
	if err != nil {
		return fmt.Errorf("serde: bind %q: %w", topic, err)
	}
	indexes := messageIndexes(prototype)
	s.mu.Lock()
	s.bindings[topic] = binding{
		prototype:     prototype,
		subject:       subject,
		schemaID:      id,
		msgIndexBytes: encodeMessageIndexes(indexes),
		indexes:       indexes,
	}
	s.mu.Unlock()
	return nil
}

// Startup binds every topic→type pair (calls Bind for each). Any failure aborts.
func (s *Serde) Startup(bindings map[string]proto.Message) error {
	for topic, prototype := range bindings {
		if err := s.Bind(topic, prototype); err != nil {
			return err
		}
	}
	return nil
}

// Produce serializes msg and wraps it in the Confluent envelope using the
// schema_id resolved at Bind. Fails if the topic was not bound.
func (s *Serde) Produce(topic string, msg proto.Message) ([]byte, error) {
	s.mu.RLock()
	b, ok := s.bindings[topic]
	s.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("serde: produce %q: %w", topic, ErrTopicNotBound)
	}

	payload, err := proto.Marshal(msg)
	if err != nil {
		return nil, fmt.Errorf("serde: produce %q: marshal: %w", topic, err)
	}

	frame := make([]byte, 0, 5+len(b.msgIndexBytes)+len(payload))
	frame = append(frame, magicByte)
	frame = binary.BigEndian.AppendUint32(frame, b.schemaID)
	frame = append(frame, b.msgIndexBytes...)
	frame = append(frame, payload...)
	return frame, nil
}

// Consume validates the Confluent envelope and deserializes into the bound type.
// Rejects (typed errors, for DLQ) when: magic byte is wrong, frame is too short,
// the schema_id is not a registered version of the topic's subject, the
// message-index does not match the bound type, or the payload fails to decode.
func (s *Serde) Consume(topic string, data []byte) (proto.Message, error) {
	s.mu.RLock()
	b, ok := s.bindings[topic]
	s.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("serde: consume %q: %w", topic, ErrTopicNotBound)
	}

	schemaID, indexes, payload, err := parseFrame(data)
	if err != nil {
		return nil, fmt.Errorf("serde: consume %q: %w", topic, err)
	}

	// Security: the id must be a registered version of THIS topic's subject —
	// not merely "exists somewhere in SR". A newer version of the same subject
	// is accepted (forward-compat); an id from another subject is rejected.
	if err := s.validateIDForSubject(schemaID, b.subject); err != nil {
		return nil, fmt.Errorf("serde: consume %q: %w", topic, err)
	}

	if !equalInts(indexes, b.indexes) {
		return nil, fmt.Errorf("serde: consume %q: %w (got %v, want %v)", topic, ErrMessageIndexMismatch, indexes, b.indexes)
	}

	msg := proto.Clone(b.prototype)
	proto.Reset(msg)
	if err := proto.Unmarshal(payload, msg); err != nil {
		return nil, fmt.Errorf("serde: consume %q: %w: %v", topic, ErrDeserialize, err)
	}
	return msg, nil
}

func topicSubject(topic string) string { return topic + "-value" }

// ── Envelope framing ────────────────────────────────────────────────────────

// parseFrame splits [magic][id_be4][msg-index][payload], validating the header.
func parseFrame(data []byte) (schemaID uint32, indexes []int, payload []byte, err error) {
	if len(data) < 6 { // magic(1) + id(4) + at least 1 index byte
		return 0, nil, nil, ErrFrameTooShort
	}
	if data[0] != magicByte {
		return 0, nil, nil, fmt.Errorf("%w (got 0x%02x)", ErrInvalidMagicByte, data[0])
	}
	schemaID = binary.BigEndian.Uint32(data[1:5])
	indexes, off, err := readMessageIndexes(data, 5)
	if err != nil {
		return 0, nil, nil, err
	}
	return schemaID, indexes, data[off:], nil
}

// messageIndexes returns the message-index path of m within its FileDescriptor.
// Truther messages are top-level, so this is a single-element path [index].
func messageIndexes(m proto.Message) []int {
	return []int{int(m.ProtoReflect().Descriptor().Index())}
}

// encodeMessageIndexes encodes the Confluent message-index array. The common
// case [0] (first message) is the 1-byte optimization 0x00.
func encodeMessageIndexes(indexes []int) []byte {
	if len(indexes) == 1 && indexes[0] == 0 {
		return []byte{0x00}
	}
	out := make([]byte, 0, (len(indexes)+1)*2)
	out = appendZigzag(out, int64(len(indexes)))
	for _, idx := range indexes {
		out = appendZigzag(out, int64(idx))
	}
	return out
}

// readMessageIndexes reads the message-index array at offset, returning the
// indexes and the offset where the payload begins.
func readMessageIndexes(data []byte, offset int) (indexes []int, next int, err error) {
	count, n, err := readZigzag(data, offset)
	if err != nil {
		return nil, 0, err
	}
	offset += n
	if count == 0 { // 1-byte optimization: single index [0]
		return []int{0}, offset, nil
	}
	indexes = make([]int, count)
	for i := 0; i < int(count); i++ {
		idx, n, err := readZigzag(data, offset)
		if err != nil {
			return nil, 0, err
		}
		indexes[i] = int(idx)
		offset += n
	}
	return indexes, offset, nil
}

func appendZigzag(b []byte, v int64) []byte {
	zz := uint64((v << 1) ^ (v >> 63))
	for zz >= 0x80 {
		b = append(b, byte(zz)|0x80)
		zz >>= 7
	}
	return append(b, byte(zz))
}

func readZigzag(data []byte, offset int) (val int64, n int, err error) {
	var ux uint64
	var shift uint
	for {
		if offset+n >= len(data) {
			return 0, 0, ErrFrameTooShort
		}
		b := data[offset+n]
		n++
		ux |= uint64(b&0x7f) << shift
		if b < 0x80 {
			break
		}
		shift += 7
	}
	return int64(ux>>1) ^ -int64(ux&1), n, nil
}

func equalInts(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// ── Schema Registry REST client (read-only) ─────────────────────────────────

// srLatestID returns the schema id of the latest registered version of subject.
func (s *Serde) srLatestID(subject string) (uint32, error) {
	endpoint := s.cfg.SRURL + "/subjects/" + url.PathEscape(subject) + "/versions/latest"
	var res struct {
		ID int `json:"id"`
	}
	if err := s.srGet(endpoint, &res); err != nil {
		return 0, err
	}
	return uint32(res.ID), nil
}

// validateIDForSubject confirms (with caching) that schema id is a registered
// version of subject, via GET /schemas/ids/{id}/versions.
func (s *Serde) validateIDForSubject(id uint32, subject string) error {
	key := fmt.Sprintf("%d|%s", id, subject)
	s.mu.RLock()
	_, ok := s.idOK[key]
	s.mu.RUnlock()
	if ok {
		return nil
	}

	endpoint := fmt.Sprintf("%s/schemas/ids/%d/versions", s.cfg.SRURL, id)
	var pairs []struct {
		Subject string `json:"subject"`
		Version int    `json:"version"`
	}
	if err := s.srGet(endpoint, &pairs); err != nil {
		// 404 → id unknown to SR entirely; srGet returns a typed not-found below.
		if errors.Is(err, errSRNotFound) {
			return fmt.Errorf("%w (schema_id=%d unknown to SR)", ErrSchemaForeign, id)
		}
		return err
	}
	for _, p := range pairs {
		if p.Subject == subject {
			s.mu.Lock()
			s.idOK[key] = struct{}{}
			s.mu.Unlock()
			return nil
		}
	}
	return fmt.Errorf("%w (schema_id=%d not under %q)", ErrSchemaForeign, id, subject)
}

var errSRNotFound = errors.New("serde: schema registry: not found")

// srGet performs an authenticated GET and decodes the JSON body into out.
func (s *Serde) srGet(endpoint string, out any) error {
	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/vnd.schemaregistry.v1+json")
	if s.cfg.APIKey != "" {
		req.SetBasicAuth(s.cfg.APIKey, s.cfg.APISecret)
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return fmt.Errorf("serde: SR GET %s: %w", endpoint, err)
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusOK:
		if err := json.NewDecoder(resp.Body).Decode(out); err != nil {
			return fmt.Errorf("serde: SR GET %s: decode: %w", endpoint, err)
		}
		return nil
	case http.StatusNotFound:
		return errSRNotFound
	default:
		b, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("serde: SR GET %s: HTTP %d: %s", endpoint, resp.StatusCode, b)
	}
}

File: sdk/node/src/serde.ts
/**
 * ProtobufSerde — thin Confluent SR serde for Kafka (Decision A).
 *
 * The SDK never reads a .proto and never registers schemas. bind() resolves the
 * topic's schema_id from the Schema Registry (read-only); produce() stamps the
 * Confluent envelope; consume() validates it (strictly) and deserializes into
 * the bound type.
 *
 * Usage:
 *   import { ProtobufSerde } from '@protobuf/contracts';
 *   import { Transaction } from '@protobuf/contracts';
 *
 *   const serde = new ProtobufSerde();             // reads SCHEMA_REGISTRY_URL
 *   await serde.bind('transactions', Transaction); // resolves schema_id at startup
 *   const framed = serde.produce('transactions', Transaction.create({ ... }));
 *   const tx = await serde.consume('transactions', rawValue); // -> Transaction
 */
import { encodeMessageIndexes, frameMessage, parseFrame, FrameError } from './framing';
import { messageIndexFor } from './message-index';
import { SchemaRegistryClient } from './schema-registry-client';
import type { MessageFns, SerdeOptions } from './types';

export type SerdeErrorCode =
  | 'INVALID_MAGIC_BYTE'
  | 'FRAME_TOO_SHORT'
  | 'TOPIC_NOT_BOUND'
  | 'SCHEMA_FOREIGN'
  | 'MESSAGE_INDEX_MISMATCH'
  | 'DESERIALIZATION_ERROR';

/**
 * Thrown by consume() on any payload rejection. The adapter routes the raw bytes
 * to the DLQ using `code` — it must never crash on a bad message.
 */
export class SerdeError extends Error {
  readonly code: SerdeErrorCode;
  readonly rawPayload?: Buffer;
  constructor(message: string, code: SerdeErrorCode, rawPayload?: Buffer) {
    super(message);
    this.name = 'SerdeError';
    this.code = code;
    this.rawPayload = rawPayload;
  }
}

interface Binding<T> {
  codec: MessageFns<T>;
  subject: string;
  schemaId: number;
  msgIndexBytes: Buffer;
  indexes: number[];
}

export class ProtobufSerde {
  private readonly sr: SchemaRegistryClient;
  private readonly bindings = new Map<string, Binding<unknown>>();

  constructor(options: SerdeOptions = {}) {
    const srUrl = options.srUrl ?? process.env['SCHEMA_REGISTRY_URL'] ?? 'http://localhost:8081';
    this.sr = new SchemaRegistryClient(
      srUrl,
      options.srApiKey ?? process.env['SCHEMA_REGISTRY_API_KEY'],
      options.srApiSecret ?? process.env['SCHEMA_REGISTRY_API_SECRET'],
    );
  }

  /**
   * Map a topic to its message type and resolve the topic's schema_id from SR
   * (subject "<topic>-value", latest version). Read-only; fails fast if the
   * subject is not registered. Call once per topic at startup.
   */
  async bind<T>(topic: string, codec: MessageFns<T>): Promise<void> {
    const subject = `${topic}-value`;
    const schemaId = await this.sr.latestId(subject);
    const indexes = messageIndexFor(codec.$type);
    this.bindings.set(topic, {
      codec: codec as MessageFns<unknown>,
      subject,
      schemaId,
      msgIndexBytes: encodeMessageIndexes(indexes),
      indexes,
    });
  }

  /** Bind every topic→codec pair (calls bind for each). */
  async startup(bindings: Record<string, MessageFns<unknown>>): Promise<void> {
    for (const [topic, codec] of Object.entries(bindings)) {
      await this.bind(topic, codec);
    }
  }

  /** Serialize msg and wrap it in the Confluent envelope (cached schema_id + correct msg-index). */
  produce<T>(topic: string, msg: T): Buffer {
    const b = this.bindings.get(topic);
    if (!b) {
      throw new SerdeError(`No binding for topic '${topic}'. Call bind() at startup.`, 'TOPIC_NOT_BOUND');
    }
    const payload = b.codec.encode(msg).finish();
    return frameMessage(b.schemaId, b.msgIndexBytes, payload);
  }

  /**
   * Validate the envelope and deserialize into the bound type. Rejects (SerdeError)
   * on: wrong magic byte, short frame, schema_id not a registered version of this
   * topic's subject, message-index ≠ bound type, or decode failure. Adapter → DLQ.
   */
  async consume<T>(topic: string, data: Buffer): Promise<T> {
    const b = this.bindings.get(topic);
    if (!b) {
      throw new SerdeError(`No binding for topic '${topic}'. Call bind() at startup.`, 'TOPIC_NOT_BOUND', data);
    }

    let schemaId: number;
    let indexes: number[];
    let payload: Buffer;
    try {
      ({ schemaId, indexes, payload } = parseFrame(data));
    } catch (err) {
      const code = err instanceof FrameError ? err.code : 'INVALID_MAGIC_BYTE';
      throw new SerdeError(`[${topic}] ${(err as Error).message}`, code, data);
    }

    // Security: the id must be a registered version of THIS topic's subject.
    if (!(await this.sr.idBelongsToSubject(schemaId, b.subject))) {
      throw new SerdeError(
        `[${topic}] schema_id=${schemaId} is not a registered version of '${b.subject}'`,
        'SCHEMA_FOREIGN',
        data,
      );
    }

    if (!indexesEqual(indexes, b.indexes)) {
      throw new SerdeError(
        `[${topic}] message-index ${JSON.stringify(indexes)} != bound ${JSON.stringify(b.indexes)}`,
        'MESSAGE_INDEX_MISMATCH',
        data,
      );
    }

    try {
      return (b.codec as MessageFns<T>).decode(payload);
    } catch (err) {
      throw new SerdeError(`[${topic}] deserialization failed: ${(err as Error).message}`, 'DESERIALIZATION_ERROR', data);
    }
  }
}

function indexesEqual(a: number[], b: number[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}
